### PR TITLE
Add ArborXWrappers::DistributedTree

### PR DESCRIPTION
**Pull request by Rombur**
This PR adds supports for `ArborX::DistributedTree`. This is the distributed version of the `ArborX::BVH` which we already have wrappers for.  With this library, we can do spatial search and nearest neighbors search when the points/bounding boxes are distributed.

### CODE REVIEW

File: include/deal.II/arborx/distributed_tree.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    template <int dim, typename Number>
+    DistributedTree(const MPI_Comm &                       comm,
+                    const std::vector<Point<dim, Number>> &points);
+
+    /**
+     * Return the indices and the MPI ranks of those BoundingBox objects that satisfy the @p queries.
+     * Because @p queries can contain multiple queries, the function returns a pair
+     * of indices and ranks and the associated offsets.
+     */
```
Comment (masterleinad) [inline include/deal.II/arborx/distributed_tree.h:60]: I think we should mention restrictions on the type/interface of `queries` here.

**Review by masterleinad — CHANGES_REQUESTED** [review]

**Comment by Rombur** [comment]
I have updated the comments

File: include/deal.II/arborx/distributed_tree.h [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+   * distributed version of ArborX:BVH.
+   *
+   * Because ArborX uses Kokkos, Kokkos needs to be initialized and finalized
+   * before using this class.
+   */
+  class DistributedTree
+  {
+  public:
+    /**
+     * Constructor. Use a vector of BoundingBox @p bounding_boxes as primitives.
```
[the remark this answers is not in the document]
Comment (Rombur) [reply include/deal.II/arborx/distributed_tree.h:42]: I'll add a comment but to answer your question, these are local bounding boxes. The way ArborX works in the distributed case is by building two trees. First it builds a "regular" tree that is only local. Then the root of the local trees are used to create a second tree that it is used to know on which processors there might be something interesting.

With the current wrappers, we return the indices and the ranks of the bounding boxes that satisfy the predicate but ArborX has a callback system that can be used to execute a function on a remote processor and get the results back. I still need to think of a nice way to expose that capability.

**Comment by Rombur** [comment]
@masterleinad good catch. It should be good now.

**Comment by Rombur** [comment]
@masterleinad ping

**Review by masterleinad — APPROVED** [review]

**Review by luca-heltai — APPROVED** [review]

**Comment by Rombur** [comment]
> Can you also add the DEAL_II_ARBORX_WITH_MPI to the doxygen file, so that the documentation shows up online?

Done

**Pull request merged**